### PR TITLE
Deprecate `TrainerCallbackHookMixin`

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?
In #10575 I removed many methods of `TrainerCallbackHookMixin`, but we should keep them for BC until v1.8 so that `Trainer.call_hook` still works as it did before, i.e. if someone calls `Trainer.call_hook("on_before_accelerator_backend_setup")` that should still work.

This PR puts back the methods removed in #10575, and then deprecates the entire `TrainerCallbackHookMixin` class.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/callback_hook.py
Comment (ananthsub) [inline pytorch_lightning/trainer/callback_hook.py:35]: Please add formal deprecation warnings along with tests so users know they'll be removed and so we remember these todos later

Comment (daniellepintz) [reply pytorch_lightning/trainer/callback_hook.py:35]: I initially did (still have it in my git stash) but then I thought it might be confusing since these hooks aren't deprecated like https://github.com/PyTorchLightning/pytorch-lightning/blob/753861a01f99543241aa4357e607ad65936c172d/pytorch_lightning/trainer/callback_hook.py#L243-L251, and I thought it might be misleading. But I can certainly add them back

Comment (daniellepintz) [reply pytorch_lightning/trainer/callback_hook.py:35]: And if users use `call_hook` they will get a deprecation warning already

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/callback_hook.py
Comment (ananthsub) [reply pytorch_lightning/trainer/callback_hook.py:35]: In the deprecation message, you can say `"`TrainerCallbackHookMixin.<hook name>` was deprecated in v1.6 and will be removed in v1.8"` since we want to communicate that the class methods are being removed.

Comment (daniellepintz) [reply pytorch_lightning/trainer/callback_hook.py]: Done. I put "removed in v1.7" since it is a mixin that really shouldn't be used by users right now, so not a typical deprecation.

File: pytorch_lightning/trainer/callback_hook.py
Comment (ananthsub) [inline pytorch_lightning/trainer/callback_hook.py:584]: are the save/load checkpoint hooks the only ones not deprecated at this point? 

I think `on_save_checkpoint` is used only here in the Trainer: https://github.com/PyTorchLightning/pytorch-lightning/blob/f95976d602d294ea3943dadd6ca4e8f2ed7bab21/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L381

And the naming is *very* confusing: if someone saw `Trainer.on_save_checkpoint` they for sure wouldn't anticipate that it only returns the callback states for callbacks attached to the trainer. especially when compared to `trainer.save_checkpoint` which saves everything. 

I think it's worth discussing (in a separate issue/PR) if we can move these functions to be within the CheckpointConnector so we can deprecate all of `TrainerCallbackHookMixin` and so the definitions here live next to the only place where they're used. 

This really only needs the callbacks, and could be a function like:
```py
get_callback_checkpoint_states(callbacks: Sequence[Callbacks]) -> Dict[str, dict]:
```

File: CHANGELOG.md
Comment (ananthsub) [inline CHANGELOG.md:182]: If the comment here holds: https://github.com/PyTorchLightning/pytorch-lightning/pull/11148#discussion_r771886148

We can later edit this to be Deprecated `TrainerCallbackHookMixin` so there's no ambiguity around which callback hooks are going to be removed

Comment (ananthsub) [inline CHANGELOG.md:182]:
```suggestion
- Deprecated most of the methods in `TrainerCallbackHookMixin` ([#11148](https://github.com/PyTorchLightning/pytorch-lightning/pull/11148))
```

File: pytorch_lightning/trainer/callback_hook.py
Comment (daniellepintz) [reply pytorch_lightning/trainer/callback_hook.py:584]: > are the save/load checkpoint hooks the only ones not deprecated at this point?

yes! Completely agree, was planning on addressing after this PR, but maybe we can address before merging this PR and then this PR can fully deprecate the `TrainerCallbackHookMixin`. Will create an issue

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by Borda** [comment]
still draft?

**Comment by daniellepintz** [comment]
@Borda I am waiting for #11179 to be merged first

**Review by ananthsub — APPROVED** [review]

**Comment by daniellepintz** [comment]
Ready for review!

**Pull request merged**
